plasmapy/formulary/distribution.py

```diff
@@ -234,28 +228,13 @@ def Maxwellian_velocity_2D(
 
 
     """
-    if units == "units":
-        # unit checks and conversions
-        # checking velocity units
-        vx = vx.to(u.m / u.s)
-        vy = vy.to(u.m / u.s)
-        # catching case where drift velocities have default values, they
-        # need to be assigned units
-        vx_drift = _v_drift_units(vx_drift)
-        vy_drift = _v_drift_units(vy_drift)
-        # convert temperature to Kelvins
-        T = T.to(u.K, equivalencies=u.temperature_energy())
-        if np.isnan(vTh):
-            # get thermal velocity and thermal velocity squared
-            vTh = parameters.thermal_speed(T, particle=particle, method="most_probable")
-        elif not np.isnan(vTh):
-            # check units of thermal velocity
-            vTh = vTh.to(u.m / u.s)
-    elif np.isnan(vTh) and units == "unitless":
-        # assuming unitless temperature is in Kelvins
-        vTh = parameters.thermal_speed(
-            T * u.K, particle=particle, method="most_probable"
-        ).si.value
+    if np.isnan(vTh):
+        # get thermal velocity and thermal velocity squared
+        vTh = parameters.thermal_speed(T, particle=particle, method="most_probable")
+    elif not np.isnan(vTh):
+        # check units of thermal velocity
+        vTh = vTh.to(u.m / u.s)
+
```

Comment:
The execution speed might get faster if, after this initial handling step, we were to drop the units like this:

```python
vTh = vTh.si.value
```

And on output, stick them back on:
```python
return u.Quantity(distFunc, u.s/u.m)
```

Another trick to speed up astropy.units is to precompute the composite units elsewhere in the code:

```python
_m_s = _velocity_unit = u.m / u.s  # name is stylistic preference
```

then refer to that precomputed unit everywhere.

Reply:
Precomputing as the velocity unit definitely helps. I've done that throughout now. 

However, adding
```
    vTh = vTh.si.value
    vx = vx.si.value
    vx_drift = vx_drift.si.value
    vy = vy.si.value
    vy_drift = vy_drift.si.value
```

after the method invocation does not really help. I believe it is the unit conversions that make things slow (that happen in the ```validate_quantities```, not so much the computations between units. 

Just to give you a number on how much slower this is: executing **all** distribution.py tests in unitless mode takes around ```4.10 seconds```. When enforcing units, executing **all but the two 3D norm tests with units** takes around ```12.55 seconds```. 

Any other ideas?

Reply:
A few ideas that may give some improvement.

1. Don't re-define the values.  I think doing `vTh = vTh.si.value` actually causes another unit check/conversion.  The `validate_quantities` has already done this, so just rely on it and use `vTh.value`.
1. This is a good case for using `pass_equivalent_units=True` in the `validate_quantities`.  This will make sure the passed unit can be converted to the expected unit but will never actually perform the conversion.  Then, just before you do your math, normalize the velocity units and cast them to dimensionless units (i.e `(vx / vTh).to(u.dimensionless_unscaled)`).  As long as the units are equivalent, there's no need to convert them ahead of time.  Doing it like this will limit the number of conversions, decrease any floating point errors due to conversions, and reduce floating point errors from doing the square [`(vx / vTh) ** 2` will be more accurate than `(vx ** 2) / (vTh ** 2)`].

Side note, I'm not going to quibble over the testing time for unit-less functions vs functions handling units.  Unit-less functions are always going to run quicker than functions handling units, that's the trade-off for having the convenience.  [*Of course I say that where the execution time is still within "reason," and we do our part to mitigate the increased execution time.*]  Additionally, if the correct units are pass then there should not be a significant difference between execution times.  Additionally, there are more test cases for functions handling units, so testing is going to take longer by default.  We really need to look at the statistics of the function executing, not the tests.

Reply:
For point 1., you are absolutely right, it does more conversions when doing ```vTh.si.value``` and simply using ```vTh.value``` speeds things up. 

I don't quite get your second point. I believe you're trying to point out to convert to dimensionless if possible. I believe this already happens, as ```np.exp``` works only with dimensionless units. I do agree that doing ```(vx / vTh) ** 2``` is better than ```(vx ** 2) / (vTh ** 2)```, but the slow cases really manifest themselves in higher dimensions where I don't think you can do the same (I might be mistaken):

```
    expTerm = np.exp(
        ((-1) * ((vx- vx_drift) ** 2 + (vy- vy_drift) ** 2) / vTh ** 2)
        .to(u.dimensionless_unscaled)
    )
    coeff = (vThSq * np.pi) ** (-1)
    distFunc = coeff * expTerm
```

I could apply this to the one dimensional cases of course and run some stats on performance.

Reply:
I'd approach it more like this...

```python
# calc exponential term
exp_xvar = ((vx - vx_drift) / vTh).to(u.dimensionless_unscaled).value
exp_yvar = ((vy - vy_drift) / vTh).to(u.dimensionless_unscaled).value
exp_term = np.exp((-1) * ((exp_xvar ** 2) + (exp_yvar ** 2)))

# calc normalization term
vTh = vTh.to(_velocity_units)
coeff = ((vTh ** 2) * n.pi) ** (-1)

return coeff * expm_term
```

Reply:
Got it, thanks. I tried it out, and no significant improvement. 

Btw., this is essentially what I'm running to test this out: https://pastebin.com/5HhST7aV

```
In [31]: %timeit run_unitless()
24.5 ms ± 1.35 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
vs. 
```
In [32]: %timeit run_units()
11.1 s ± 559 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
Whoa, that's 3 orders of magnitude of a difference...

Reply:
Looking over your tests I understand your timing issues better.  I didn't realize you were running an integration test, I thought it was just the package tests.  This does bring up an issue that I have definitely been sweeping under the rug.  These functions are inherently going to be integrated and there are not a lot of unit friendly integration routines out there.  Thus, I can definitely see the benefit of a unit-less version.

We could possibly do something like...

```python
def Maxellian_1D_unitless(v, ...):
    # do stuff
    return distribution

Maxwellian_1D = validate_quantities(v=_velocity_units)(Maxwellian_1D_unitless)
```

The main issue with this approach is documentation.

But now my inner @namurphy is screaming out "ABSTRACT BASE CLASSES" and we should convert everything to a class based structure instead.  Then we could add a builtin integrator.

Sorry @FilipBolt, I'm definitly diverging off the original path.  I think there's a lot to be discussed here, but my primary issue for this PR is how the arguments are setup (see my comments bellow...https://github.com/PlasmaPy/PlasmaPy/pull/816#discussion_r432645641 and https://github.com/PlasmaPy/PlasmaPy/pull/816#discussion_r432673709).

Reply:
I could see a potential class structure looking like...

```python
class MaxwellianDistribution:
    def __init__(self, vTh: _velocity_units = None, particle='e', order=1, type='velocity'):
        self._vTh = vTh
        self._particle = particle
        self._T = None
        self._order = order
        self._type = type

        if vTh is None:
            warn("not temperature or thermal velocity set")

    @property
    def temperature(self) -> u.K:
        return self._T

    @temperature.setter
    def temperature(self, val: u.K, particle=None)
        # do all your conditioning, unit validations, calc vTh, etc.

    # make properties & setters for vTh and particle

    def _probability_density_unitless(self, v):
        # no units involved, assumed to be correct
        # do calculations
        return prob_dens

    @validate_quantities
    def probability_density(self, v: _velocity_units):
        # this would be unit flavored
        # this is analogous to our current functions
        # do your unit conditioning and call _probability_density_unitless()
        # add units back on
        return prob_dens

    @validate_quantities
    def probability(self, vstart: _velocity_units, vstop: _velocity_units):
        # this would perform the integral from vstart to vstop
        # condition the units and strip
        # integrate using _probability_density_unitless
        return prob
```

use would be like...

```python
>>> dist = MaxwellianDistribution()
>>> dist.temperature = 100 * u.K
>>> dist.probability_density(20 * u.m / u.s)
some value
>>> dist.probability(20 * u.m / u.s, 40 * u.m / u.s)
some value
```

Reply:
Ok, does this mean we are not comfortable with letting go of units for the sake of only calculating ```probability``` or in general? 

Everyone that does probability via integration themselves would need to migrate to a new interface to have it as efficient as now. 

I completely understand your concerns @rocco8773 and they I would say it is vital to have them resolved to before discussing a solution for a way forward.